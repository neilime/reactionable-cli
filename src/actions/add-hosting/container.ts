import 'reflect-metadata';

import { Container } from 'inversify';

import Amplify from './adapters/amplify/Amplify';
import { HostingAdapter } from './adapters/HostingAdapter';
import Netlify from './adapters/netlify/Netlify';

export const AdapterKey = 'HostingAdapter';

export function bindHostingAdapters(container: Container): void {
  container.bind<HostingAdapter>(AdapterKey).to(Amplify);
  container.bind<HostingAdapter>(AdapterKey).to(Netlify);
}
