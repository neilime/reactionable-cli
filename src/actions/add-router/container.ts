import "reflect-metadata";

import { Container } from "inversify";

import NextJs from "./adapters/NextJs";
import { RouterAdapter } from "./adapters/RouterAdapter";
import RouterDom from "./adapters/RouterDom";

export const AdapterKey = "RouterAdapter";

export function bindRouterAdapters(container: Container): void {
  container.bind<RouterAdapter>(AdapterKey).to(NextJs);
  container.bind<RouterAdapter>(AdapterKey).to(RouterDom);
}
