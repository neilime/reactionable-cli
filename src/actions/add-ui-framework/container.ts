import 'reflect-metadata';

import { Container } from 'inversify';

import { IUIFrameworkAdapter } from './adapters/IUIFrameworkAdapter';
import ReactBootstrap from './adapters/ReactBootstrap';

export const AdapterKey = 'UIFrameworkAdapter';

export function bindUIFrameworkAdapters(container: Container) {
  container.bind<IUIFrameworkAdapter>(AdapterKey).to(ReactBootstrap);
}
