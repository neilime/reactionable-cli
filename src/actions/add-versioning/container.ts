import 'reflect-metadata';
import { Container } from 'inversify';
import Git from './adapters/github/Github';
import { IVersioningAdapter } from './IVersioningAdapter';

let container = new Container({ defaultScope: 'Singleton' });

container.bind<IVersioningAdapter>('Adapter').to(Git);

export default container;