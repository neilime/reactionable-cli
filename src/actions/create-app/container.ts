import "reflect-metadata";

import { Container } from "inversify";

import CreateNextApp from "./adapters/create-next-app/CreateNextApp";
import CreateReactApp from "./adapters/create-react-app/CreateReactApp";
import { CreateAppAdapter } from "./adapters/CreateAppAdapter";

export const AdapterKey = "CreateAppAdapter";

export function bindCreateAppAdapters(container: Container): void {
  container.bind<CreateAppAdapter>(AdapterKey).to(CreateReactApp);
  container.bind<CreateReactApp>(CreateReactApp).toSelf();
  container.bind<CreateAppAdapter>(AdapterKey).to(CreateNextApp);
  container.bind<CreateNextApp>(CreateNextApp).toSelf();
}
