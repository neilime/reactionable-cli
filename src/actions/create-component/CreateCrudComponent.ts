import { prompt } from 'inquirer';
import { inject, injectable } from 'inversify';
import { plural, singular } from 'pluralize';

import { ConsoleService } from '../../services/ConsoleService';
import { FileService } from '../../services/file/FileService';
import { PackageManagerService } from '../../services/package-manager/PackageManagerService';
import { TemplateService } from '../../services/TemplateService';
import AddHosting from '../add-hosting/AddHosting';
import AddUIFramework from '../add-ui-framework/AddUIFramework';
import CreateComponent, { CreateComponentOptions } from './CreateComponent';

@injectable()
export default class CreateCrudComponent extends CreateComponent {
  constructor(
    @inject(AddUIFramework) addUIFramework: AddUIFramework,
    @inject(AddHosting) addHosting: AddHosting,
    @inject(PackageManagerService)
    protected readonly packageManagerService: PackageManagerService,
    @inject(ConsoleService) protected readonly consoleService: ConsoleService,
    @inject(FileService) protected readonly fileService: FileService,
    @inject(TemplateService) protected readonly templateService: TemplateService
  ) {
    super(
      addUIFramework,
      addHosting,
      packageManagerService,
      consoleService,
      fileService,
      templateService
    );
  }

  getName(): string {
    return 'Create a new react CRUD component';
  }

  async run({ realpath, name }: CreateComponentOptions): Promise<void> {
    if (!name) {
      const answer = await prompt<{ name: string }>([
        {
          name: 'name',
          message: "What's the component entity name?",
          validate: (input) => (input.length ? true : 'Component entity name is required'),
        },
      ]);
      name = answer.name;
    }
    const entityName = this.formatName(name);
    const entitiesName = plural(entityName);
    const templateContext = { entityName, entitiesName };
    this.consoleService.info(`Create CRUD component for "${entityName}"...`);

    if (!this.fileService.fileDirExistsSync(realpath)) {
      throw new Error(
        `Unable to create CRUD component for "${entityName}" in unexisting directory "${realpath}`
      );
    }

    // Create main component
    const componentDirPath = await this.createComponent({
      realpath,
      name: plural(entityName),
      componentTemplate: 'crud/Crud.tsx',
      templateContext,
    });

    // Create config
    const configTemplateContext = {
      ...templateContext,
      uiPackage: await this.getUIPackage(realpath),
    };
    await this.templateService.renderTemplateTree(
      componentDirPath,
      CreateComponent.templateNamespace + '/crud',
      { [entitiesName + 'Config.tsx']: 'Config.tsx' },
      configTemplateContext
    );

    // Create child components
    const components = {
      Create: entityName,
      Delete: entityName,
      Update: entityName,
      Read: entityName,
      List: entitiesName,
    };

    for (const componentName of Object.keys(components)) {
      await this.createComponent({
        realpath,
        componentDirPath,
        name: componentName + components[componentName],
        componentTemplate: `crud/${componentName.toLowerCase()}/${componentName}.tsx`,
        templateContext,
      });
    }

    this.consoleService.success(`CRUD component for "${entityName}" has been created`);
  }

  protected formatName(name: string): string {
    return singular(super.formatName(name));
  }
}
