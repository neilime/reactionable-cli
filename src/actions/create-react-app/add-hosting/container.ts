import 'reflect-metadata';
import { Container } from 'inversify';
import { IHostingAction } from './hostings/IHostingAction';
import Amplify from './hostings/amplify/Amplify';

let container = new Container({ defaultScope: 'Singleton' });

container.bind<IHostingAction>('Hosting').to(Amplify);

export default container;