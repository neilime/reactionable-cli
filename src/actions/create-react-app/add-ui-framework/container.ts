import 'reflect-metadata';
import { Container } from 'inversify';
import ReactBootstrap from './ui-frameworks/ReactBootstrap';
import { IUIFrameworkAction } from './ui-frameworks/IUIFrameworkAction';

let container = new Container({ defaultScope: 'Singleton' });

container.bind<IUIFrameworkAction>('UIFramework').to(ReactBootstrap);

export default container;