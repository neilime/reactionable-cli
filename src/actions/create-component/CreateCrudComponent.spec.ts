import { stat, readFile } from "fs/promises";
import { resolve } from "path";

import prompts from "prompts";

import container from "../../container";
import { DirResult, createTmpDir } from "../../tests/tmp-dir";
import CreateCrudComponent from "./CreateCrudComponent";

describe("createCrudComponent", () => {
  let createCrudComponent: CreateCrudComponent;

  let testDir: DirResult;

  beforeAll(async () => {
    createCrudComponent = container.get(CreateCrudComponent);
  });

  describe("construct", () => {
    it("should be initialized", () => {
      expect(createCrudComponent).toBeInstanceOf(CreateCrudComponent);
    });
  });

  describe.each([
    ["React", "test-react-project", "src"],
    ["NextJs", "test-nextjs-project", "lib"],
  ])("Run for a %s project", (name, testProjectPath, libPath) => {
    let testDirPath: string;

    beforeAll(async () => {
      testDir = await createTmpDir(testProjectPath);
      testDirPath = testDir.name;

      prompts.inject(["overwrite"]);
      await createCrudComponent.run({
        realpath: testDirPath,
        name: "test entity",
      });
    });

    afterAll(() => testDir?.removeCallback());

    const expectedFiles = [
      // Config
      "components/test-entities/TestEntitiesConfig.tsx",
      // Crud entry point
      "components/test-entities/TestEntities.test.tsx",
      "components/test-entities/TestEntities.tsx",
      // Create
      "components/test-entities/create-test-entity/CreateTestEntity.test.tsx",
      "components/test-entities/create-test-entity/CreateTestEntity.tsx",
      // Read
      "components/test-entities/read-test-entity/ReadTestEntity.test.tsx",
      "components/test-entities/read-test-entity/ReadTestEntity.tsx",
      // Update
      "components/test-entities/update-test-entity/UpdateTestEntity.test.tsx",
      "components/test-entities/update-test-entity/UpdateTestEntity.tsx",
      // Delete
      "components/test-entities/delete-test-entity/DeleteTestEntity.test.tsx",
      "components/test-entities/delete-test-entity/DeleteTestEntity.tsx",
      // List
      "components/test-entities/list-test-entities/ListTestEntities.test.tsx",
      "components/test-entities/list-test-entities/ListTestEntities.tsx",
      // Translations
      "i18n/i18n.ts",
      "i18n/locales/en/testEntities.json",
      "i18n/locales/fr/testEntities.json",
    ];

    it.each(expectedFiles.map((value) => [value]))(
      'should create crud component file "%s"',
      async (expectedFile) => {
        const filePath = resolve(testDirPath, libPath, expectedFile);

        expect((await stat(filePath)).isFile()).toBe(true);
        expect(await readFile(filePath, "utf-8")).toMatchSnapshot(expectedFile);
      }
    );
  });
});
