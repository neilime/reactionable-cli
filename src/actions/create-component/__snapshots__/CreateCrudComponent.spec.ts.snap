// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCrudComponent run 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router';
import CreateTestEntity from './CreateTestEntity';

it('renders without crashing', () => {  
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><CreateTestEntity /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
});"
`;

exports[`CreateCrudComponent run 2`] = `
"import { Create, IFormProps, IOnSubmitForm } from '@reactionable/ui-bootstrap';
import React from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { ITestEntityData, ITestEntityValues, testEntityFormChildren, testEntityFormSchema, testEntityFormValues } from '../TestEntitiesConfig';

interface IProps extends Pick<IFormProps<ITestEntityValues, ITestEntityData>, 'onSuccess'> { }
const CreateTestEntity: React.FC<IProps> = (props) => {
  const { t } = useTranslation();
  const onSubmit: IOnSubmitForm<ITestEntityValues, ITestEntityData> = async (values, formikHelpers) => {
    return { id: '', label: '' };
  };

  const form = {
    title: t('Create a new test entity'),
    onSubmit,
    formValues: testEntityFormValues,
    formSchema: testEntityFormSchema(t),
    chilren: testEntityFormChildren(t),
    ...props,
  };

  return <>
    
    <Create<ITestEntityValues, ITestEntityData> form={form} />
  </>;
};
export default CreateTestEntity;"
`;

exports[`CreateCrudComponent run 3`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router';
import DeleteTestEntity from './DeleteTestEntity';

it('renders without crashing', () => {  
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><DeleteTestEntity /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
});"
`;

exports[`CreateCrudComponent run 4`] = `
"import { Delete, IDeleteProps } from '@reactionable/ui-bootstrap';
import React from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { ITestEntityData } from '../TestEntitiesConfig';

interface IProps extends Pick<IDeleteProps<ITestEntityData>, 'onSuccess'> {
  id: string; 
  label?: boolean;
}
const DeleteTestEntity: React.FC<IProps> = (props) => {
  const { t } = useTranslation();
  const { id, label, ...deleteProps} = props;
  const onConfirm = async (): Promise<ITestEntityData> => {
    return { id, label: '' };
  };

  return <>
    
    <Delete<ITestEntityData>
      {...deleteProps}
      title={t('Delete test entity')}
      children={label ? t('Delete test entity') : undefined}
      confirmationMessage={t('Are you sure you want to delete this test entity ?')}
      successMessage={t('The test entity has been deleted')}
      onConfirm={onConfirm}
    />
  </>;
};
export default DeleteTestEntity;"
`;

exports[`CreateCrudComponent run 5`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router';
import ListTestEntities from './ListTestEntities';

it('renders without crashing', () => {  
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><ListTestEntities /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
});"
`;

exports[`CreateCrudComponent run 6`] = `
"import { useQueryList } from '@reactionable/amplify';
import { List } from '@reactionable/ui-bootstrap';
import React, { lazy } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { Link, generatePath, useRouteMatch } from 'react-router-dom';
import { IListTestEntitiesQueryVariables, ITestEntityData } from '../TestEntitiesConfig';

const CreateTestEntity = lazy(() => import('../create-test-entity/CreateTestEntity'));
const DeleteTestEntity = lazy(() => import('../delete-test-entity/DeleteTestEntity'));
interface IProps { }
const ListTestEntities: React.FC<IProps> = (props) => {
  const { t } = useTranslation();
  const match = useRouteMatch();

  const { data, refetch, next, previous, ...listProps } = useQueryList<ITestEntityData, IListTestEntitiesQueryVariables>({
  });

  const renderChildren = (data: ITestEntityData) => <tr key={'TestEntity-' + data.id}>
    <td>
      <Link
        to={generatePath(\`\${match.path}/:testEntitiesId\`, { ...match.params, testEntitiesId: data.id })}
        title={t('Go to test entity \\"{{label}}\\"', data)}
      >{data.label}</Link>
    </td>
    <td>
      <Link
        to={generatePath(\`\${match.path}/:testEntitiesId\`, { ...match.params, testEntitiesId: data.id })}
        title={t('Go to test entity \\"{{label}}\\"', data)}
      >{t('Go to test entity \\"{{label}}\\"', data)}</Link>
      <DeleteTestEntity id={data.id} onSuccess={refetch}>
        {t('Delete test entity \\"{{label}}\\"', data)}
      </DeleteTestEntity>
    </td>
  </tr>;

  return <>
    <Helmet>
      <title>{t('tests')} - {t('List test entities')}</title>
      <meta name=\\"description\\" content={t('List test entities description')} />
    </Helmet>
    <h1>{t('List test entities')}</h1>
    <Link disabled={!previous} title={t('Load the previous test entities')}>{t('Previous')}</Link>
    <Link disabled={!next} title={t('Load the next test entities')}>{t('Next')}</Link>
    <List<ITestEntityData>
      {...listProps}
      data={data}
      head={[t('Label'), t('Actions')]}
      children={renderChildren}
      noData={<>
        <br /> <br />
        {t('You don\\\\'t have any test entities yet')}
        <br />
        {t('Click the button to create a new test entity')}
        <br /><br />
        <CreateTestEntity onSuccess={refetch}>
          <Link>{t('Create a new test entity')}</Link>
        </CreateTestEntity>
      </>}
    />
  </>;
};
export default ListTestEntities;"
`;

exports[`CreateCrudComponent run 7`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router';
import ReadTestEntity from './ReadTestEntity';

it('renders without crashing', () => {  
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><ReadTestEntity /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
});"
`;

exports[`CreateCrudComponent run 8`] = `
"import { Read } from '@reactionable/ui-bootstrap';
import React, { lazy } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { Link, useParams } from 'react-router-dom';
import { IGetTestEntityQueryVariables, ITestEntityData } from '../TestEntitiesConfig';

const UpdateTestEntity = lazy(() => import('../update-test-entity/UpdateTestEntity'));
interface IProps { }
const ReadTestEntity: React.FC<IProps> = (props) => {
  const { t } = useTranslation();
  const matchParams = useParams<{ testEntityId: string }>();

  const { data, refetch, ...readProps } = useQuery<ITestEntityData, IGetTestEntityQueryVariables>({
    variables: { id: matchParams.testEntityId },
  });

  const renderChildren = (data: ITestEntityData) => <>
    <Helmet>
      <title>{t('tests')} - {t('Read test entity - {{label}}', data)}</title>
      <meta name=\\"description\\" content={t('Read test entity - {{label}}', data)} />
    </Helmet>
    <h1>{t('Read test entity - {{label}}', data)}</h1>
    { data && <>
      <UpdateTestEntity
        onSuccess={refetch}
        formValues={data}
      >
        <Link to=\\"#\\">{ t('Update test entity \\"{{label}}\\"', data) }</Link>
      </UpdateTestEntity>
    </>}
  </>;

  return <>
    
    <Read<ITestEntityData>
      {...readProps}
      data={data}
      children={renderChildren}
    />
  </>;
};
export default ReadTestEntity;"
`;

exports[`CreateCrudComponent run 9`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router';
import UpdateTestEntity from './UpdateTestEntity';

it('renders without crashing', () => {  
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><UpdateTestEntity /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
});"
`;

exports[`CreateCrudComponent run 10`] = `
"import { IFormProps, IOnSubmitForm, Update } from '@reactionable/ui-bootstrap';
import React from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { useParams } from 'react-router-dom';
import { ITestEntityData, ITestEntityValues, testEntityFormChildren, testEntityFormSchema, testEntityFormValues } from '../TestEntitiesConfig';

interface IProps extends Pick<IUpdateProps<ITestEntityValues, ITestEntityData>, 'onSuccess' | 'formValues'> { }
const UpdateTestEntity: React.FC<IProps> = (props) => {
  const { t } = useTranslation();
  const matchParams = useParams<{ testEntityId: string }>();
  const onSubmit: IOnSubmitForm<ITestEntityValues, ITestEntityData> = async (values, formikHelpers) => {
    return { id: '', label: '' };
  };

  const safeFormValues: any = { ...testEntityFormValues };
  Object.keys(safeFormValues).forEach(value => {
    const key = value as keyof ITestEntityValues;
    if (props.formValues[key]) {
      safeFormValues[key] = props.formValues[key] as ITestEntityValues[keyof ITestEntityValues];
    }
  });

  const form = {
    title: t('Update test entity \\"{{label}}\\"', safeFormValues),
    onSubmit,
    formValues: safeFormValues,
    formSchema: testEntityFormSchema(t),
    chilren: testEntityFormChildren(t),
    ...props,
  };

  return <>
    
    <Update<ITestEntityValues, ITestEntityData> form={form} />
  </>;
};
export default UpdateTestEntity;"
`;

exports[`CreateCrudComponent run 11`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router';
import TestEntities from './TestEntities';

it('renders without crashing', () => {  
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><TestEntities /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
});"
`;

exports[`CreateCrudComponent run 12`] = `
"import { Crud } from '@reactionable/core';
import { IUseLayoutProps } from '@reactionable/ui-bootstrap';
import React, { lazy } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { ITestEntityData } from './TestEntitiesConfig';

const ListTestEntities = lazy(() => import('./list-test-entities/ListTestEntities'));
const ReadTestEntity = lazy(() => import('./read-test-entity/ReadTestEntity'));
interface IProps { }
const TestEntities: React.FC<IProps> = (props) => {
  const { t } = useTranslation();
  
  return <>
    
    <Crud<ITestEntityData, IUseLayoutProps>
      name=\\"TestEntity\\"
      listComponent={ ListTestEntities }
      readComponent={ ReadTestEntity }
      {...props}
    />
  </>;
};
export default TestEntities;"
`;

exports[`CreateCrudComponent run 13`] = `
"import { IFormChildrenProps } from '@reactionable/core';
import { FormField } from '@reactionable/ui-bootstrap';
import { TFunction } from 'i18next';
import React from 'react';
import { string } from 'yup';

export type ITestEntityData = {
  id: string;
  label: string;
};
export type ITestEntityValues = Omit<ITestEntityData, 'id'>;
export type IGetTestEntityQueryVariables = Pick<ITestEntityData, 'id'> & {

};
export type IListTestEntitiesQueryVariables = {
  
};
export const testEntityFormValues: ITestEntityValues = {
  label: '',
};
export const testEntityFormSchema = (t: TFunction) => ({
   label: string()
    .min(1, t('Test entity label must be at least 1 characters'))
    .required(t('Test entity label is required')),
});
export const testEntityFormChildren = (t: TFunction) => (formikProps: IFormChildrenProps<ITestEntityValues>) => <>
  <FormField
    name=\\"label\\"
    label={t('Label')}
    autoFocus={true}
  />
</>;"
`;
